java/monkey/src/main/java/root/REPL.java

```diff
@@ -0,0 +1,31 @@
+package root;
+
+import java.io.InputStream;
+import java.io.PrintStream;
+import java.util.Scanner;
+
+public class REPL {
+
+    private static final String PROMPT = ">> ";
+
+    public void start(InputStream inputStream, PrintStream printStream) {
+        var scanner = new Scanner(inputStream);
+
+        while (true) {
```

Comment:
I'd swap some stuff around here.

The biggest issue is that you're declaring `while (true)`, but also have a break inside of the loop. This means _we do not loop forever_.

I believe a much better result would be a `do-while` loop, where `line.trim().length() == 0` sets a boolean to `false`, to end the loop.

Reply:
I don't really thing a while (true) with a break is bad, it seems like a pretty common pattern to me. But I have two ideas to improve it:

```java
var scanner = new Scanner(inputStream);
var finished = false;

do {
    printStream.print(PROMPT);
    var line = scanner.nextLine();

    if (line.trim().length() != 0) {
        var lexer = new Lexer(line);

        var t = lexer.nextToken();
        while (t.type() != TokenType.EOF) {
            printStream.println(t);
            t = lexer.nextToken();
        }
    } else {
        finished = true;
    }
} while (!finished);
```
and
```java
var scanner = new Scanner(inputStream);

printStream.print(PROMPT);
var line = scanner.nextLine();

while (line.trim().length() != 0) {
    var lexer = new Lexer(line);

    var t = lexer.nextToken();
    while (t.type() != TokenType.EOF) {
        printStream.println(t);
        t = lexer.nextToken();
    }

    printStream.print(PROMPT);
    line = scanner.nextLine();
}
```
But i do believe the original one to be more clear.

Reply:
You could also make a wrapper class.
```java
public class LexerIterator implements Iterator<Token> {
    private final Lexer lexer;
    public LexerIterator(Lexer lexer) {
        this.lexer = lexer;
    }
    public static LexerIterator fromString(String input) {
       return new LexerIterator(new Lexer(input));
    }
    
    @Override
    public boolean hasNext() {
        return lexer.peek() != '\0';
    }

    @Override
    public Token next() {
        return lexer.nextToken();
    }
}
```
If you wanted to create a Token Buffer instead to match against Token Types, that would be fine. Actually might be better.

With this, we can adjust to the following
```java
printStream.print(PROMPT);
while (scanner.hasNextLine()) { // Might need some extra handling here for the initial pass, not sure
   var line = scanner.nextLine();
   LexerIterator.fromString(line).forEachRemaining(this::printToken); 
   printStream.print(PROMPT);
} // The exception would break us out of the loop.

public void printToken(Token token) {
     if (token.type() == TokenType.ILLEGAL) {
        throw new IllegalArgumentException("Illegal Token Provided", token.literal());
     }
     System.out.println(token);
}
```
Easier to read the flow of the logic, and we get the option to treat the Lexer as a forward looking Iterator.

Reply:
`while(true)` may be "common", but it's a ban common. It implies that you never leave the context of the loop, a better pattern would be this.
```java
boolean exitLoop = false;
while (!exitLoop) {
     if (exitCondition) {
        exitLoop = true;
        continue;
     }
}
```
`do/while` being more preferred.